File: packages/aethyme/src/scorecard/detectors/ability_coverage.py
"""Detector for permission/ability coverage."""

import re
from pathlib import Path
from typing import TypedDict

from ..models import Finding, Severity
from .base import BaseDetector


class ProtectedRoute(TypedDict):
    file: Path
    line: int
    content: str


class AbilityCoverageDetector(BaseDetector):
    """Detect missing permission checks and ability definitions."""

    @property
    def name(self) -> str:
        return "ability-coverage"

    @property
    def description(self) -> str:
        return "Checks for missing authorization and permission definitions"

    def detect(self) -> list[Finding]:
        """Detect missing ability/permission checks."""
        findings: list[Finding] = []

        # Find protected routes
        routes = self._find_protected_routes()

        # Check for permission definitions
        findings.extend(self._check_permission_definitions(routes))

        # Check for authorization checks
        findings.extend(self._check_authorization_checks())

        return findings

    def _find_protected_routes(self) -> list[ProtectedRoute]:
        """Find routes that should have permission checks."""
        routes: list[ProtectedRoute] = []

        # Patterns for routes with auth
        auth_patterns = [
            r'@router\.(post|put|delete|patch)',  # Mutating operations
            r'Depends\(get_current_user\)',  # Auth dependency
            r'@require_auth',
            r'@login_required',
        ]

        for file_path in self.repo_path.rglob('*.py'):
            if self.should_skip_file(file_path):
                continue

            content = self.read_file_safe(file_path)
            if not content:
                continue

            lines = content.split('\n')

            for line_num, line in enumerate(lines, start=1):
                for pattern in auth_patterns:
                    if re.search(pattern, line):
                        routes.append({
                            'file': file_path,
                            'line': line_num,
                            'content': line.strip()
                        })
                        break

        return routes

    def _check_permission_definitions(self, routes: list[ProtectedRoute]) -> list[Finding]:
        """Check if routes have permission checks."""
        findings: list[Finding] = []

        permission_patterns = [
            r'can\(',
            r'check_permission',
            r'require_permission',
            r'@permission_required',
            r'has_permission',
            r'authorize\(',
        ]

        for route in routes:
            file_path = route["file"]
            line_num = route["line"]

            content = self.read_file_safe(file_path)
            if not content:
                continue

            lines = content.split('\n')

            # Check surrounding lines for permission checks
            has_permission = False

            start = max(0, line_num - 5)
            end = min(len(lines), line_num + 15)  # Check more lines after

            for i in range(start, end):
                line = lines[i] if i < len(lines) else ""

                for pattern in permission_patterns:
                    if re.search(pattern, line):
                        has_permission = True
                        break

                if has_permission:
                    break

            if not has_permission:
                findings.append(Finding(
                    detector=self.name,
                    severity=Severity.WARNING,
                    message="Protected route missing explicit permission check",
                    file_path=str(file_path.relative_to(self.repo_path)),
                    line_number=line_num,
                    evidence=route["content"][:100],
                    suggestion="Add explicit permission check, e.g., check_permission('resource:action')"
                ))

        return findings

    def _check_authorization_checks(self) -> list[Finding]:
        """Check for authorization model definitions."""
        findings: list[Finding] = []

        # Look for ability/permission definition files
        ability_files = list(self.repo_path.glob('**/abilities.py')) + \
                       list(self.repo_path.glob('**/permissions.py')) + \
                       list(self.repo_path.glob('**/authorization.py'))

        if not ability_files:
            # No centralized ability definitions found
            findings.append(Finding(
                detector=self.name,
                severity=Severity.INFO,
                message="No centralized permission/ability definitions found",
                file_path=".",
                line_number=None,
                evidence=None,
                suggestion="Create abilities.py or permissions.py to centralize authorization logic"
            ))

        return findings

File: packages/aethyme/rust/crates/aethyme-quality/src/detectors/ability_coverage.rs
//! Detector for permission/ability coverage (port of
//! `src/scorecard/detectors/ability_coverage.py`).

use std::path::{Path, PathBuf};
use std::sync::OnceLock;

use regex::Regex;

use crate::model::{Finding, Severity};
use crate::util::{py_slice, py_strip};
use crate::walk::{any_entry_named, read_file_safe, rglob_all, should_skip_file};

pub struct AbilityCoverageDetector;

struct ProtectedRoute {
    file: PathBuf,
    line: usize,
    content: String,
}

struct Patterns {
    auth: Vec<Regex>,
    permission: Vec<Regex>,
}

fn patterns() -> &'static Patterns {
    static CELL: OnceLock<Patterns> = OnceLock::new();
    CELL.get_or_init(|| Patterns {
        auth: vec![
            Regex::new(r"@router\.(post|put|delete|patch)").unwrap(),
            Regex::new(r"Depends\(get_current_user\)").unwrap(),
            Regex::new(r"@require_auth").unwrap(),
            Regex::new(r"@login_required").unwrap(),
        ],
        permission: vec![
            Regex::new(r"can\(").unwrap(),
            Regex::new(r"check_permission").unwrap(),
            Regex::new(r"require_permission").unwrap(),
            Regex::new(r"@permission_required").unwrap(),
            Regex::new(r"has_permission").unwrap(),
            Regex::new(r"authorize\(").unwrap(),
        ],
    })
}

impl super::Detector for AbilityCoverageDetector {
    fn name(&self) -> &'static str {
        "ability-coverage"
    }

    fn description(&self) -> &'static str {
        "Checks for missing authorization and permission definitions"
    }

    fn detect(&self, repo_path: &Path) -> Vec<Finding> {
        let routes = self.find_protected_routes(repo_path);
        let mut findings = self.check_permission_definitions(repo_path, &routes);
        findings.extend(self.check_authorization_checks(repo_path));
        findings
    }
}

impl AbilityCoverageDetector {
    fn find_protected_routes(&self, repo_path: &Path) -> Vec<ProtectedRoute> {
        let patterns = patterns();
        let mut routes = Vec::new();

        for entry in rglob_all(repo_path) {
            // rglob('*.py'): fnmatch NAME filter (see schema_drift's
            // note on the directory-named-x.py edge).
            let is_py = entry
                .path
                .file_name()
                .map(|n| n.to_string_lossy().ends_with(".py"))
                .unwrap_or(false);
            if !is_py {
                continue;
            }
            if should_skip_file(&entry.path) {
                continue;
            }
            let Some(content) = read_file_safe(&entry.path) else {
                continue;
            };
            if content.is_empty() {
                continue;
            }
            for (line_num, line) in content.split('\n').enumerate() {
                let line_num = line_num + 1;
                if patterns.auth.iter().any(|p| p.is_match(line)) {
                    routes.push(ProtectedRoute {
                        file: entry.path.clone(),
                        line: line_num,
                        content: py_strip(line).to_string(),
                    });
                }
            }
        }

        routes
    }

    fn check_permission_definitions(
        &self,
        repo_path: &Path,
        routes: &[ProtectedRoute],
    ) -> Vec<Finding> {
        let patterns = patterns();
        let mut findings = Vec::new();

        for route in routes {
            // Python re-read the file per route.
            let Some(content) = read_file_safe(&route.file) else {
                continue;
            };
            if content.is_empty() {
                continue;
            }
            let lines: Vec<&str> = content.split('\n').collect();
            let line_num = route.line; // 1-based

            // Window: 0-based indices [line_num-5, line_num+15) — 4
            // lines above through 14 below, decorator line included.
            let start = line_num.saturating_sub(5);
            let end = (line_num + 15).min(lines.len());
            let has_permission = (start..end).any(|i| {
                let line = lines.get(i).copied().unwrap_or("");
                patterns.permission.iter().any(|p| p.is_match(line))
            });

            if !has_permission {
                let rel = route
                    .file
                    .strip_prefix(repo_path)
                    .unwrap_or(&route.file)
                    .to_string_lossy()
                    .to_string();
                findings.push(Finding {
                    detector: "ability-coverage".to_string(),
                    severity: Severity::Warning,
                    message: "Protected route missing explicit permission check".to_string(),
                    file_path: rel,
                    line_number: Some(line_num as i64),
                    evidence: Some(py_slice(&route.content, 100)),
                    suggestion: Some(
                        "Add explicit permission check, e.g., check_permission('resource:action')"
                            .to_string(),
                    ),
                });
            }
        }

        findings
    }

    /// Port of `_check_authorization_checks`: pure existence globs
    /// (`**/abilities.py` etc.) with NO skip filter — a match under
    /// node_modules or .git counts, exactly like the Python glob.
    fn check_authorization_checks(&self, repo_path: &Path) -> Vec<Finding> {
        let any_ability_file = any_entry_named(repo_path, "abilities.py")
            || any_entry_named(repo_path, "permissions.py")
            || any_entry_named(repo_path, "authorization.py");
        if any_ability_file {
            return Vec::new();
        }
        vec![Finding {
            detector: "ability-coverage".to_string(),
            severity: Severity::Info,
            message: "No centralized permission/ability definitions found".to_string(),
            file_path: ".".to_string(),
            line_number: None,
            evidence: None,
            suggestion: Some(
                "Create abilities.py or permissions.py to centralize authorization logic"
                    .to_string(),
            ),
        }]
    }
}

#[cfg(test)]
mod tests {
    use super::super::Detector;
    use super::*;
    use crate::testsupport::{build_good_scorecard_repo, build_problematic_scorecard_repo, tmpdir};

    // Translated from tests/scorecard/test_detectors.py::TestAbilityCoverageDetector.

    #[test]
    fn detects_missing_permissions() {
        let tmp = tmpdir("ability-bad");
        let repo = build_problematic_scorecard_repo(&tmp);
        let findings = AbilityCoverageDetector.detect(&repo);
        assert!(!findings.is_empty());
    }

    #[test]
    fn good_repo_checks_pass() {
        let tmp = tmpdir("ability-good");
        let repo = build_good_scorecard_repo(&tmp);
        // permissions.py exists and the route window contains
        // check_permission → no findings at all.
        assert!(AbilityCoverageDetector.detect(&repo).is_empty());
    }

    #[test]
    fn exact_shape_on_problematic_repo() {
        let tmp = tmpdir("ability-shape");
        let repo = build_problematic_scorecard_repo(&tmp);
        let findings = AbilityCoverageDetector.detect(&repo);
        // Two protected-route hits on the same file (Depends(get_current_user)
        // appears on the decorated def line, @router.post above it), plus
        // the no-central-definitions info finding.
        assert_eq!(findings.len(), 3);
        assert!(
            findings[..2]
                .iter()
                .all(|f| f.message == "Protected route missing explicit permission check"
                    && f.severity == Severity::Warning
                    && f.file_path == "src/api/routes.py")
        );
        let info = &findings[2];
        assert_eq!(info.severity, Severity::Info);
        assert_eq!(info.file_path, ".");
        assert_eq!(
            info.message,
            "No centralized permission/ability definitions found"
        );
    }

    #[test]
    fn ability_files_anywhere_suppress_info_finding() {
        let tmp = tmpdir("ability-glob");
        let repo = tmp.join("r");
        // Even under node_modules (no skip filter on the glob).
        let nested = repo.join("node_modules/pkg");
        std::fs::create_dir_all(&nested).unwrap();
        std::fs::write(nested.join("permissions.py"), "x = 1\n").unwrap();
        assert!(AbilityCoverageDetector.detect(&repo).is_empty());
    }
}
